enhancements/ebpf-in-ocp.md

```diff
@@ -0,0 +1,211 @@
+---
+
+# Guidelines for the Use of eBPF in OCP
+
+## Summary
+
+OCP developers and customers alike would like to use eBPF in OCP
+clusters. At the present time, _some_ eBPF hooks are designed to be
+"shareable", but others are not (and some are shareable only when used
+in certain restricted ways). We need to create rules for how OCP will
+use eBPF and how we expect others to use (or not use it) to prevent
+conflicts later.
+
+## Motivation
+
+### User Stories
+
+#### Existing Use Case Number 1
+
+(Something about observability or tracing where people are already
+using eBPF in OCP.)
+
+#### Upcoming Network Observability Features
+
+(Something about upcoming network observability features.)
+
+#### Ensuring Network (and General) Supportability
+
+As an OCP developer or support engineer, I would like to be able to
+debug customer problems without having to worry that random
+"unsanctioned" eBPF is messing things up in inexplicable ways.
+
+#### Customer eBPF Bug Patching
+
+As an OCP customer, I want to work around an OCP/RHEL bug by using a
+`MachineConfig` (or something similar) to deploy a small eBPF program
+to every node in my cluster, without voiding my warranty.
+
+I might be OK with needing a support exception, since this is just to
+work around a bug that should eventually be fixed.
+
+#### Customer eBPF Feature Patching
+
+As an OCP customer, I want to add functionality to OCP/RHEL by using a
+`MachineConfig` (or something similar) to deploy a small (or
+not-so-small) eBPF program to every node in my cluster, without
+voiding my warranty.
+
+I might be OK with needing a support exception.
+
+(eg, concretely, there is a customer who wants to use eBPF to
+implement [OVS-style SLB
+Bonding](https://docs.openvswitch.org/en/latest/topics/bonding/#slb-bonding)
+on a kernel bond interface, using eBPF. The upstream Linux maintainer
+rejected a patch to implement it in the kernel driver directly, in
+part because it would be so easy to implement it with eBPF.)
+
+#### Cilium as a Supported Third-Party Network Plugin
+
+As an OCP customer, I want to deploy an OCP cluster using Cilium as a
+certified third-party network plugin, without worrying that any other
+OCP components are going to interfere with it (or vice versa).
+
+Cilium uses a wide variety of eBPF hooks, including FIXME.
+
+#### eBPF From OLM
+
+As a third-party developer of an operator that uses eBPF, I would like
+to make my operator available via OLM and have it interoperate
+smoothly with existing and future OCP releases.
+
+#### Future eBPF Data Plane Functionality
+
+As an OCP developer/product manager, I would like to introduce
+functionality in a future release that makes use of the XDP and/or tc
+eBPF hooks to modify data plane functionality, and have these new
+features interact in a predictable way with existing networking
+functionality, other operators, and local customizations.
+
+#### Security
+
+As a cluster administrator, I would like to make sure that no one is
```

Comment:
Currently only privileged pods would be able to run ebpf programs... So the admin should know about any of these right?  Also As we start running bpf programs it probably will make sense to have an OCP system pod that gives some stats on all of the bpf programs currently running in a cluster (maybe this is observability's Job)

Reply:
Being more precise, it is not mandatory that pods are full-privileged to run eBPF programs. For example, the NetObserv eBPF agent runs in a non-privileged pod, in a container defining the following security context:

```yaml
        securityContext:
          capabilities:
            add:
              - BPF
              - PERFMON
              - NET_ADMIN
              - SYS_RESOURCE
          runAsUser: 0
```

Reply:
Would SELinux profile (CAP_BPF) help keep the attack surface a little smaller than using privileged containers (when needed)?